### PR TITLE
Leader election pre-requisites. 

### PR DESCRIPTION
**Pull request by SUMUKHA-PK**
This PR lays a foundation upon which we can build leader election.

### CODE REVIEW

File: servermanagement/routing/startRaft.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ func StartRaft(w http.ResponseWriter, r *http.Request) {
 		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
-	fmt.Printf("I %s am a %s\n", r.Host, newReq[r.Host].State)
+	fmt.Printf("I %s, am a %s\n", r.Host, newReq[r.Host].ServerState.Name)
```
Comment (abhishekkumar2718) [inline servermanagement/routing/startRaft.go:29]: Please use log instead of fmt

File: servermanagement/startSignal.go [hunk 2]
```diff
@@ -15,13 +15,24 @@ import (
 func StartSignal(config types.Configuration, RaftServers map[string]types.RaftServer) error {
 	wg := &sync.WaitGroup{}
 	wg.Add(len(config.Servers))
-	designationMap := make(map[string]string)
+	designationMap := make(map[string]types.State)
+	state := types.State{
```
Comment (abhishekkumar2718) [inline servermanagement/startSignal.go]: `types.State` will likely have many more functions related to. Let's create a new file for it and write a function `BuildInitialState()`

File: types/types.go [hunk 3]
```diff
@@ -12,19 +12,56 @@ type Configuration struct {
...
+	ServerState State  `json:"serverState"`
+	ID          int    `json:"id"`
+	IP          string `json:"IP"`
+	Port        string `json:"Port"`
 	// the designation map holds the mapping to
 	// the IP of the server to the state of the
 	// server in the cluster. This is useful to
 	// send out data to necessary servers.
-	DesignationMap map[string]string `json:"designationMap"`
+	DesignationMap map[string]State `json:"designationMap"`
```
Comment (abhishekkumar2718) [inline types/types.go]: Let's call it simply `ServerDesignations` or `URIDesignations` instead.

File: types/types.go [hunk 4]
```diff
@@ -12,19 +12,56 @@ type Configuration struct {
...
 	// dont hear from a leader, they can become
 	// candidates. Leaders are elected from the
 	// leader election process.
-	State string `json:"state"`
-	IP    string `json:"IP"`
-	Port  string `json:"Port"`
+	Name        string
+	CurrentTerm int
+	// VotedFor maintains the ID of the voted
+	// server; -1 if its leader, -2 at init
```
Comment (abhishekkumar2718) [inline types/types.go:32]: 1. What's the advantage of using -1 instead of self ID?

File: types/types.go [hunk 5]
```diff
@@ -12,19 +12,56 @@ type Configuration struct {
...
+	// that is known to be replicated on the server
+	MatchIndex []int
+	// above 2 variables are volatile only int the
+	// leader and for each follower. Its also
+	// re-init after each election.
+}
+
+// RaftServer describes a single raft server
+type RaftServer struct {
+	ServerState State  `json:"serverState"`
```
Comment (abhishekkumar2718) [inline types/types.go]: Let's rename this to `CurrentState` instead.

File: types/types.go (hunk 4 above)
Comment (SUMUKHA-PK) [reply types/types.go:32]: Spec says use null. So I followed it.

File: types/types.go (hunk 3 above)
Comment (abhishekkumar2718) [inline types/types.go]: Do we need to store the complete state or just whether the node is a follower, candidate or leader? There are a lot of details in state which will be useless - Log, VotedFor etc.

File: servermanagement/startSignal.go [hunk 6]
```diff
@@ -3,50 +3,51 @@ package servermanagement
...
 // according to the raft protocol.
 func StartSignal(config types.Configuration, RaftServers map[string]types.RaftServer) error {
-	wg := &sync.WaitGroup{}
-	wg.Add(len(config.Servers))
-	designationMap := make(map[string]string)
+	designationMap := make(map[string]types.State)
 	for i := range config.Servers {
-		designationMap[config.Servers[i].IP+":"+config.Servers[i].Port] = "follower"
+		designationMap[config.Servers[i].IP+":"+config.Servers[i].Port] = types.State{
+			Name:        "follower",
```
Comment (abhishekkumar2718) [inline servermanagement/startSignal.go]: Name follower? It should rather be designation.

If we are storing names, I will add another field (But there were no names in the type.Server).

Honestly, we don't need to store designation either. We can compute it using VotedFor.

File: servermanagement/routing/concurrentReqRes.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+					return err
+				}
+				req.Header.Add("Content-Type", "application/json")
+
+				res, err := http.DefaultClient.Do(req)
+				if err != nil {
+					log.Printf("Bad response in startSignal.go: %v\n", err)
+					return err
+				}
+				defer res.Body.Close()
```
Comment (abhishekkumar2718) [inline servermanagement/routing/concurrentReqRes.go:35]: Just curious about this, if the function returns at the error check above. Will the deferred function be called?

File: servermanagement/startSignal.go (hunk 6 above)
Comment (SUMUKHA-PK) [reply servermanagement/startSignal.go]: Its in the `State` struct. State.Name is follower.

File: servermanagement/routing/concurrentReqRes.go (hunk 7 above)
Comment (SUMUKHA-PK) [reply servermanagement/routing/concurrentReqRes.go:35]: We'll have to move defer up.

Comment (SUMUKHA-PK) [reply servermanagement/routing/concurrentReqRes.go:35]: Um actually we _have_ to check for errors before `defer` apparently.

File: servermanagement/startSignal.go (hunk 6 above)
Comment (SUMUKHA-PK) [reply servermanagement/startSignal.go]: Sure, we can remove it. It was just used for checks and tests till now.

**Comment by SUMUKHA-PK** [comment]
Removing `DesignationMap` its not needed after the recent changes.

**Comment by SUMUKHA-PK** [comment]
Merging this. Functionally approved.

**Pull request merged**